db/db.go

```diff
@@ -245,9 +239,289 @@ func (db *DB) execImpl(basis types.Ref, function string, args types.List) (newDa
+	defer transactionMutex.Unlock()
+	txID := transactionCounter
+	transactionCounter++
+	tx := &readTransaction{txID, db.noms, db.head}
+	transactions[txID] = tx
+	return tx
+}
+
+func (db *DB) ReadTransaction(body func(tx ReadTransaction)) error {
+	tx := db.BeginReadTransaction()
+	body(tx)
+	return tx.Close()
+}
+
+func (db *DB) CloseTransaction(txID int) error {
+	tx, err := db.FindReadTransaction(txID)
+	if err != nil {
+		return err
+	}
+	return tx.Close()
+}
+
+func (db *DB) CommitTransaction(txID int) (types.Ref, error) {
+	tx, err := db.FindWriteTransaction(txID)
+	if err != nil {
+		return types.Ref{}, err
+	}
+	return tx.Commit()
+}
+
+func (db *DB) AbortTransaction(txID int) error {
+	tx, err := db.FindWriteTransaction(txID)
+	if err != nil {
+		return err
+	}
+	return tx.Abort()
+}
+
+func (db *DB) BeginWriteTransaction() WriteTransaction {
+	transactionMutex.Lock()
+	defer transactionMutex.Unlock()
+	txID := transactionCounter
+	transactionCounter++
+	tx := &writeTransaction{txID, db, db.noms, db.head, db.head.Data(db.noms).Edit(), false}
+	transactions[txID] = tx
+	return tx
+}
+
+func (db *DB) WriteTransaction(body func(tx WriteTransaction)) error {
+	tx := db.BeginWriteTransaction()
+	body(tx)
+	_, err := tx.Commit()
+	return err
+}
+
 func (db *DB) lock() func() {
 	db.mu.Lock()
 	return func() {
 		db.mu.Unlock()
 	}
 }
+
+// ReadTransaction describes the operations allowed in a read transaction.
+type ReadTransaction interface {
+	Get(id string) ([]byte, error)
+	Has(id string) (bool, error)
+	Scan(opts ScanOptions) ([]ScanItem, error)
+	ID() int
+	Close() error
+	Closed() bool
+}
+
+// WriteTransaction describes the operations allowed in a write transaction.
+type WriteTransaction interface {
+	ReadTransaction
+	Put(path string, JSON []byte) error
+	Del(path string) (ok bool, err error)
+	Commit() (types.Ref, error)
+	Abort() error
+}
+
+type readTransaction struct {
+	id   int
+	noms datas.Database
+	head Commit
+}
+
+func (tx *readTransaction) ID() int {
+	return tx.id
+}
+
+func (tx *readTransaction) Closed() bool {
+	return tx.id == -1
+}
+
+func (tx *readTransaction) kvMap() kv.Map {
+	return tx.head.Data(tx.noms)
+}
+
+func (tx *readTransaction) Get(id string) ([]byte, error) {
+	if tx.Closed() {
+		return nil, ClosedError
+	}
+	value := tx.kvMap().Get(types.String(id))
+	if value == nil {
+		return nil, nil
+	}
+	var b bytes.Buffer
+	err := nomsjson.ToJSON(value, &b)
+	return b.Bytes(), err
+}
+
+func (tx *readTransaction) Has(id string) (bool, error) {
+	if tx.Closed() {
+		return false, ClosedError
+	}
+	return tx.kvMap().Has(types.String(id)), nil
+}
+
+func (tx *readTransaction) Scan(opts ScanOptions) ([]ScanItem, error) {
+	if tx.Closed() {
+		return nil, ClosedError
+	}
+	// TODO fritz clean up
+	return scan(tx.kvMap().NomsMap(), opts)
+}
+
+func (tx *readTransaction) Close() error {
+	if tx.Closed() {
+		return ClosedError
+	}
+	tx.id = -1
+	removeTransaction(tx)
+	return nil
+}
+
+type writeTransaction struct {
```

Comment:
Why draw the distinction in the Go code between read and write transactions? I think they can all be write transactions and it can just be an API/typesystem thing whether it's a read or write tx.

I refreshed my memory and the cost of a `MapEditor` with no edits is basically zero. So there's no reason to avoid eagerly creating one.

It seems like you could eliminate half the code by literally just deleting the `readTransaction` type and wiring the callers to `writeTransaction`.

Reply:
Scan is not yet implemented on a `MapEditor`

Reply:
I'll look into doing the iterator on MapEditor before continuing.

Reply:
I think the simplest solution is to just build the Map at that point. We do not want to reflect changes to the MapEditor during scan anyway. If this ever becomes a perf issue we can implement a `types.MapIterator` or a smaller subset of that interface for `types.MapEditor`

Reply:
It's true that `MapIterator` lacks scan, but I don't see the relevance to the question of whether to combine these two structs. Both `ReadTransaction` and `WriteTransaction` require `Scan`.